### PR TITLE
add multilabel feature

### PR DESCRIPTION
**Pull request by flyseaworld**
支持多标签训练

### CODE REVIEW

File: configs/quick_start/ResNet50_vd_multilabel.yaml [hunk 1]
```diff
@@ -0,0 +1,80 @@
+mode: 'train'
+ARCHITECTURE:
+    name: 'ResNet50_vd'
+
+pretrained_model: "./pretrained/ResNet50_vd_10w_pretrained"
```
Comment (littletomatodonkey) [inline configs/quick_start/ResNet50_vd_multilabel.yaml]: 上面添加一个注释，说明一下该pretrain model的下载地址吧，因为首页上没有提供这个，或者换成ResNet50_vd_ssld的模型也可以（动态图模型的话，下面的load_static_weights就可以去掉了）

File: configs/quick_start/ResNet50_vd_multilabel.yaml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    params:
+        momentum: 0.9
+    regularizer:
+        function: 'L2'
+        factor: 0.000070
+
+TRAIN:
+    batch_size: 64
+    num_workers: 4
+    file_list: "/path/to/multi_train_list_example.txt"
```
Comment (littletomatodonkey) [inline configs/quick_start/ResNet50_vd_multilabel.yaml]: 这块可以是否方便在高级教程中添加文档说明呢？提供一个开源的demo数据集，然后怎么训练评估预测和导出：
https://github.com/PaddlePaddle/PaddleClas/tree/develop/docs/zh_CN/advanced_tutorials

File: ppcls/data/reader.py [hunk 3]
```diff
@@ -193,6 +193,36 @@ def __len__(self):
...
+        self.full_lines = get_file_list(params)
+        self.delimiter = params.get("delimiter", "\t")
+        self.ops = create_operators(params["transforms"])
+        self.num_samples = len(self.full_lines)
+        return
+
+    def __getitem__(self, idx):
+        line = self.full_lines[idx]
+        img_path, label_str = line.split(self.delimiter)
+        img_path = os.path.join(self.params["data_dir"], img_path)
```
Comment (littletomatodonkey) [inline ppcls/data/reader.py]: 这块建议参考下develop分支的Dataset代码，添加了异常处理的代码

File: ppcls/modeling/loss.py [hunk 4]
```diff
@@ -42,6 +42,19 @@ def _labelsmoothing(self, target):
         soft_target = paddle.reshape(soft_target, shape=[-1, self._class_dim])
         return soft_target
 
+    def _binary_crossentropy(self, input, target, use_pure_fp16=False):
+        if self._label_smoothing:
+            target = self._labelsmoothing(target)
+            cost = F.binary_cross_entropy_with_logits(logit=input, label=target)
+        else:
+            cost = F.binary_cross_entropy_with_logits(logit=input, label=target)
+        if use_pure_fp16:
```
Comment (littletomatodonkey) [inline ppcls/modeling/loss.py]: 这块可以精简一下，目前use_pure_fp16是否为true，使用的都是mean，可以把这个参数去掉

File: tools/program.py [hunk 5]
```diff
@@ -131,27 +140,40 @@ def create_metric(out,
...
+        top1 = paddle.metric.accuracy(softmax_out, label=label, k=1)
+        # set topk to fetchs
+        k = min(topk, classes_num)
+        topk = paddle.metric.accuracy(softmax_out, label=label, k=k)
+
+        # multi cards' eval
+        if mode != "train" and paddle.distributed.get_world_size() > 1:
+            top1 = paddle.distributed.all_reduce(
+                top1, op=paddle.distributed.ReduceOp.
+                SUM) / paddle.distributed.get_world.size()
```
Comment (littletomatodonkey) [inline tools/program.py]: get_world_size 这里api写错了

File: tools/program.py [hunk 6]
```diff
@@ -131,27 +140,40 @@ def create_metric(out,
...
+        out = F.sigmoid(out)
+        accuracys = multilabel_metrics(out, label)
+        
+        # multi cards' eval
+        if mode != "train" and paddle.distributed.get_world_size() > 1:
+            accuracys = paddle.distributed.all_reduce(
+                accuracys, op=paddle.distributed.ReduceOp.
+                SUM) / paddle.distributed.get_world_size()
+
+        fetchs["multilabel_accuracys"] = to_tensor(accuracys)
```
Comment (littletomatodonkey) [inline tools/program.py]: 可以把paddle.distributed.all_reduc的操作抽出来，不管是否为multi label，拿到的都是order dict，再遍历一下做下all reduce即可，不用再每个判断中做

File: tools/program.py [hunk 7]
```diff
@@ -286,13 +337,17 @@ def run(dataloader,
...
+        if not multilabel:
+            topk_name = 'top{}'.format(config.topk)
+            metric_list.insert(
+                0, (topk_name, AverageMeter(
+                    topk_name, '.5f', postfix=",")))
+            metric_list.insert(
+                0, ("top1", AverageMeter(
+                    "top1", '.5f', postfix=",")))
+        else:
+            metric_list.insert(0, ("multilabel_accuracys", AverageMeter(
```
Comment (littletomatodonkey) [inline tools/program.py]: 这个要不修改为multilabel_accuracy吧，

File: tools/program.py [hunk 8]
```diff
@@ -373,4 +428,7 @@ def run(dataloader,
 
     # return top1_acc in order to save the best model
     if mode == 'valid':
-        return metric_list['top1'].avg
+        if multilabel:
+            return metric_list['multilabel_accuracys']
```
Comment (littletomatodonkey) [inline tools/program.py]: 同上

**Review by cuicheng01 — DISMISSED** [review]

**Review by cuicheng01 — DISMISSED** [review]
